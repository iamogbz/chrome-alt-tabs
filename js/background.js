/*
 *  Document    : background.js
 *  Author      : ogbizi
 *  Description : Chrome Tab Genuis background script
 *  Dependencies: chrome api
 *  Domain      : com.chrometabgenius
 */
"use strict";
var CMD_CUT = "tabs-cut";
var CMD_COPY = "tabs-copy";
var CMD_PASTE = "tabs-paste";
var CMD_DUP = "tabs-duplicate";
var CMD_UNDO = "tabs-paste-undo";
var CMD_MOVE_NEXT = "tabs-move-next";
var CMD_MOVE_PREV = "tabs-move-prev";

var app = function() {
    // command to be executed
    var pendingCommand = undefined;
    // array of tabs to be pasted
    var selectedTabs = [];
    // last command executed
    var lastCommand = undefined;
    // window tabs were pasted from
    var sourceWindow = undefined;
    // array of tabs just pasted
    var pastedTabs = [];
    // get tab id from tab objects
    var getTabIds = function (tabs) {
        var tabIds = tabs.map(function (t){ return t.id; });
        console.log("tab ids: ", tabIds);
        return tabIds;
    };
    
    return {
        setSourceWindow: function(window) {
            console.log('set source window:', window);
            sourceWindow = window;
        },
        setPendingCommand: function(command) {
            console.log('set pending command:', command);
            pendingCommand = command;
        },
        getPendingCommand: function() {
            return pendingCommand;
        },
        setSelectedTabs: function(tabs) {
            console.log("set selected tabs:", tabs);
            selectedTabs = tabs;
        },
        getSelectedTabs: function() {
            return selectedTabs;
        },
        pasteTabs: function(window, tabs, command) {
            console.log("paste tabs:", window, tabs, command);
            lastCommand = command;
            pendingCommand = undefined;
            var wId = window == undefined ? undefined : window.id;
            // perform copy/move function
            if(command == CMD_CUT) {
                pastedTabs = [];
                chrome.tabs.move(getTabIds(tabs), {windowId: wId, index:-1}, 
                    function (_tabs) {
                        // save pasted tabs for undo action
                        if(_tabs.length == undefined) {
                            pastedTabs.push(_tabs);
                        } else pastedTabs = _tabs;
                        console.log("moved tabs: ", pastedTabs);
                        // set first tab to active and highlight all
                        chrome.tabs.update(pastedTabs[0].id, {active:true});
                        pastedTabs.map(function(t) {
                            chrome.tabs.update(t.id, {highlighted:true});
                        });
                    });
            } else if(command == CMD_COPY) {
                // TODO use duplication and then move
                pastedTabs = []
                for(var i = 0; i < tabs.length; i++) {
                    // set first tab to active and highlight all
                    chrome.tabs.create({windowId: wId, url: tabs[i].url, 
                        active:true}, 
                        function(tab) {
                            pastedTabs.push(tab);
                            chrome.tabs.update(tab.id, {highlighted:true});
                        });
                }
                console.log("copied tabs: ", pastedTabs);
            }
            // switch to window
            chrome.windows.update(wId, {focused:true});
        },
        undoLastCommand: function() {
            if(lastCommand == undefined || pastedTabs.length == 0) {
                return false;
            } else {
                // switch to source window
                chrome.windows.update(sourceWindow.id, {focused:true});
                if(lastCommand == CMD_COPY) {
                    chrome.tabs.remove(getTabIds(pastedTabs), function() {
                        console.log("removed tabs: ", pastedTabs);
                        pastedTabs = [];
                        sourceWindow = undefined;
                    });
                } else if(lastCommand == CMD_CUT) {
                    chrome.tabs.move(getTabIds(pastedTabs), 
                        {windowId: sourceWindow.id, index:-1}, 
                        function (_tabs) {
                            console.log("unmoved tabs: ", _tabs);
                            // set first tab to active and highlight all
                            chrome.tabs.update(_tabs[0].id, {active:true});
                            _tabs.map(function(t) {
                                chrome.tabs.update(t.id, {highlighted:true});
                            });
                        });
                        pastedTabs = [];
                        sourceWindow = undefined;
                }
                return true;
            }
        }
    };
}();

document.addEventListener('DOMContentLoaded', function() {
    console.log("DOM loaded");
});

chrome.commands.onCommand.addListener(function(command) {
    console.log('command listener:', command);
    switch(command) {
        /*case CMD_CUT:
        case CMD_COPY:
            app.setPendingCommand(command);
            chrome.windows.getLastFocused({}, function (window) {
                console.log("last focused window:", window);
                app.setSourceWindow(window);
                chrome.tabs.query({highlighted: true, windowId: window.id}, app.setSelectedTabs);
            });
            break;
        case CMD_DUP:
            app.setPendingCommand(command);
            chrome.windows.getLastFocused({}, function (window) {
                console.log("duplication window:", window);
                app.setSourceWindow(window);
                chrome.tabs.query({highlighted: true, windowId: window.id}, function(tabs){
                    app.setSelectedTabs(tabs);
                    app.pasteTabs(undefined, tabs, CMD_COPY);
                });
            });
            break;
        case CMD_PASTE:
            chrome.windows.getLastFocused({}, function (window) {
                console.log("last focused window:", window);
                app.pasteTabs(window, app.getSelectedTabs(), app.getPendingCommand());
            });
            break;*/
        case CMD_UNDO:
            if(app.undoLastCommand()) { 
                console.log("undo successful");
            } else console.log("undo failed");
            break;
        case CMD_CUT:
        case CMD_MOVE_NEXT:
        case CMD_MOVE_PREV:
            app.setPendingCommand(command);
            chrome.windows.getLastFocused({}, function (window) {
                console.log("move from window:", window);
                app.setSourceWindow(window);
                chrome.tabs.query({highlighted: true, windowId: window.id}, function(tabs){
                    app.setSelectedTabs(tabs);
                    chrome.windows.getAll({windowTypes: ['normal']}, function (windows){
                        if(windows.length == 1 || command == CMD_CUT) {
                            var tabA = tabs[0];
                            chrome.windows.create({tabId: tabA.id}, 
                                function (nWindow){
                                    app.pasteTabs(nWindow, tabs, CMD_CUT);
                                });
                        } else {
                            var idx = 0;
                            do {
                                if(windows[idx].id == window.id) {
                                    idx += (command == CMD_MOVE_NEXT ? 1 : -1);
                                    break;
                                }
                            } while(++idx < windows.length);
                            if(idx == windows.length) idx = 0;
                            else if(idx < 0) idx = windows.length - 1;
                            var nWindow = windows[idx];
                            app.pasteTabs(nWindow, tabs, CMD_CUT);
                        }
                    });
                });
            });
            break;
    }
});