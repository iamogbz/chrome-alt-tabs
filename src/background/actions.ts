export const MOVE_TABS = "move-tabs";
export const UNDO = "undo";

export const moveTabs = ({
    tabs,
    from,
    to = null,
}: {
    tabs: ChromeTab[];
    from?: number;
    to?: number;
}): Action => {
    if (!tabs || !tabs.length || !from) {
        throw new Error(`Invalid Action Definition: move ${tabs} from ${from}`);
    }
    return {
        type: MOVE_TABS,
        payload: {
            from,
            tabs: tabs.map((t: ChromeTab) => ({ ...t })),
            to,
        },
    };
};

/**
 * Create action to undo last
 */
export const undo = (): Action => ({ type: UNDO });
