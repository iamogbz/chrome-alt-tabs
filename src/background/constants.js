const prefixCommand = command => `tabs-move-${command}`;
const commandKeys = ["out", "back", "next", "prev"];
export const COMMANDS = commandKeys.reduce(
    (commands, key) =>
        Object.assign(commands, { [key.toUpperCase()]: prefixCommand(key) }),
    {},
);

export const UNDO_LIMIT = 10;
