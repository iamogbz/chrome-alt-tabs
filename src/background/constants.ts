const prefixCommand = (command: string): string => `tabs-move-${command}`;
const commandKeys = ["out", "back", "next", "prev"];
export const COMMANDS: { [key: string]: string } = commandKeys.reduce(
  (commands, key) => ({ ...commands, [key.toUpperCase()]: prefixCommand(key) }),
  {} as Record<string, string>,
);

export const UNDO_LIMIT = 10;
