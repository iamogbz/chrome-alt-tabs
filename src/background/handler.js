import {
    createWindow,
    focusOnTab,
    focusOnWindow,
    moveTabsToWindow,
    selectTabs,
} from "../utils/chrome";
import { UNDO_LIMIT } from "./constants";
import { MOVE_TABS, UNDO } from "./actions";

const noop = () => {};
const actionLog = [];
const actionHandlers = {};

/**
 * Handle move action
 * @param {{}} action
 * @returns {Promise<Boolean>}
 */
// TODO: save tab indexes
const doMove = async ({ tabs, to }) => {
    const tabIds = [];
    let activeTab = null;
    tabs.forEach(({ id, active }) => {
        if (active) activeTab = id;
        tabIds.push(id);
    });
    const windowId = to || (await createWindow(activeTab)).id;
    await moveTabsToWindow(tabIds, windowId).catch(e => console.warn(e));
    if (activeTab) await focusOnTab(activeTab);
    await selectTabs(tabIds);
    await focusOnWindow(windowId);
    return true;
};

/**
 * Undo the move action
 * @param {{}} action
 * @returns {Promise<Boolean>}
 */
doMove.undo = async ({ from, tabs, to }) => {
    await doMove({ from: to, tabs, to: from });
    return false;
};

/**
 * Undo last action in the log
 */
const undo = async () => {
    const lastAction = actionLog.pop();
    const handler = (lastAction && actionHandlers[lastAction.type]) || noop;
    if (typeof handler.undo === "function") {
        await handler.undo(lastAction);
    }
    return false;
};

// Map action types to handlers
Object.assign(actionHandlers, {
    [MOVE_TABS]: doMove,
    [UNDO]: undo,
});

/**
 * Execute action and store in log for undo if possible
 * @param {[{}]} action
 */
export default async function(action) {
    if (!action) return;
    const canUndo = await (actionHandlers[action.type] || noop)(action);
    if (canUndo) actionLog.push(action);
    if (actionLog.length > UNDO_LIMIT) actionLog.shift();
}
