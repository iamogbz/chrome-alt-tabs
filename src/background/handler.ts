import { log } from "../utils/base";
import {
    createWindow,
    focusOnTab,
    focusOnWindow,
    moveTabsToWindow,
    selectTabs,
} from "../utils/chrome";
import { UNDO_LIMIT } from "./constants";
import { MOVE_TABS, UNDO, moveTabs } from "./actions";

const actionLog: Action[] = [];
const noActionOp = async (_: Action) => false;
const doActionHandlers: { [type: string]: ActionHandler } = {};
const undoActionHandlers: { [type: string]: ActionHandler } = {};

/**
 * Handle move action
 */
doActionHandlers[MOVE_TABS] = async ({
    payload: { tabs, to },
}: Action): Promise<boolean> => {
    const tabIds: number[] = [];
    let activeTab: number = null;
    tabs.forEach(({ id, active }: ChromeTab) => {
        if (active) activeTab = id;
        tabIds.push(id);
    });
    const windowId: number = to || (await createWindow(activeTab)).id;
    await moveTabsToWindow(tabs, windowId).catch(e => log.error(e));
    await focusOnTab(activeTab || tabIds[0]);
    await selectTabs(tabIds);
    await focusOnWindow(windowId);
    return true;
};

/**
 * Undo the move action
 */
undoActionHandlers[MOVE_TABS] = async ({
    payload: { from, tabs },
}: Action): Promise<boolean> => {
    await doActionHandlers[MOVE_TABS](moveTabs({ tabs, to: from }));
    return false;
};

/**
 * Undo last action in the log
 */
doActionHandlers[UNDO] = async () => {
    const lastAction = actionLog.pop() || ({} as Action);
    const undoHandler = undoActionHandlers[lastAction.type] || noActionOp;
    await undoHandler(lastAction);
    return false;
};

/**
 * Execute action and store in log for undo if possible
 */
export default async function(action: Action): Promise<void> {
    if (!action) return;
    const canUndo = await (doActionHandlers[action.type] || noActionOp)(action);
    if (canUndo) actionLog.push(action);
    if (actionLog.length > UNDO_LIMIT) actionLog.shift();
}
