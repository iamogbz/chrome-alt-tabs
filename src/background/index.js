import {
    onCommand,
    getLastFocusedWindow,
    getSelectedTabsInWindow,
    getWindowAfter,
    getWindowBefore,
} from "../utils/chrome";
import { COMMANDS } from "./constants";
import { moveTabs, undo } from "./actions";
import handle from "./handler";

/**
 * Get the context of the current command activated
 * @returns {Promise}
 */
const getCommandContext = async () => {
    const currentWindow = await getLastFocusedWindow();
    const selectedTabs = await getSelectedTabsInWindow(currentWindow.id);
    return { currentWindow, selectedTabs };
};

/**
 * Wrap a function to be executed with the command context when activated
 * @param {Function} fn to be called with the command context
 */
const withCommandContext = async fn => {
    const {
        currentWindow: { id: windowId },
        selectedTabs,
    } = await getCommandContext();
    await fn({ windowId, selectedTabs });
};

const commandActions = [
    [
        COMMANDS.OUT,
        withCommandContext(({ windowId: from, selectedTabs: tabs }) => {
            handle(moveTabs({ tabs, from }));
        }),
    ],
    [
        COMMANDS.NEXT,
        withCommandContext(async ({ windowId: from, selectedTabs: tabs }) => {
            const { id: to } = await getWindowAfter(from);
            handle(moveTabs({ tabs, from, to }));
        }),
    ],
    [
        COMMANDS.PREV,
        withCommandContext(async ({ windowId: from, selectedTabs: tabs }) => {
            const { id: to } = await getWindowBefore(from);
            handle(moveTabs({ tabs, from, to }));
        }),
    ],
    [COMMANDS.UNDO, () => handle(undo())],
];
commandActions.forEach(args => onCommand(...args));
