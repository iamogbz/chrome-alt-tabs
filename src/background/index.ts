import "crx-livereload";

import {
  getAllTabsInWindow,
  getLastFocusedWindow,
  getSelectedTabsInWindow,
  getWindowIdAfter,
  getWindowIdBefore,
  onCommand,
} from "utils/chrome";
import { moveTabs, undo } from "./actions";
import { COMMANDS } from "./constants";
import { handleAction } from "./handler";

/**
 * Get the context of the current command activated
 * @returns {Promise}
 */
const getCommandContext = async (): Promise<{
  currentWindow: ChromeWindow;
  selectedTabs: ChromeTab[];
  allTabs: ChromeTab[];
}> => {
  const currentWindow = await getLastFocusedWindow();
  const [allTabs, selectedTabs] = await Promise.all([
    getAllTabsInWindow(currentWindow.id),
    getSelectedTabsInWindow(currentWindow.id),
  ]);
  return { currentWindow, selectedTabs, allTabs };
};

/**
 * Wrap a function to be executed with the command context when activated
 * @param {Function} fn to be called with the command context
 */
const withCommandContext = (fn: (...args: unknown[]) => void) => async () => {
  const { currentWindow, selectedTabs, allTabs } = await getCommandContext();
  const isAllTabsSelected = selectedTabs.length === allTabs.length;
  return fn({ windowId: currentWindow.id, selectedTabs, isAllTabsSelected });
};

export const commandActions = {
  [COMMANDS.OUT]: withCommandContext(
    async ({ windowId, selectedTabs: tabs, isAllTabsSelected }) => {
      if (isAllTabsSelected) return;
      await handleAction(moveTabs({ tabs, from: windowId }));
    }
  ),
  [COMMANDS.NEXT]: withCommandContext(
    async ({ windowId, selectedTabs: tabs, isAllTabsSelected }) => {
      const from = isAllTabsSelected ? null : (windowId as number);
      const to = await getWindowIdAfter(from);
      await handleAction(moveTabs({ tabs, from, to }));
    }
  ),
  [COMMANDS.PREV]: withCommandContext(
    async ({ windowId, selectedTabs: tabs, isAllTabsSelected }) => {
      const from = isAllTabsSelected ? null : (windowId as number);
      const to = await getWindowIdBefore(from);
      await handleAction(moveTabs({ tabs, from, to }));
    }
  ),
  [COMMANDS.BACK]: async (): Promise<void> => {
    await handleAction(undo());
  },
};
const noop = (): void => undefined;
export const commandListener = (command: string): void =>
  (commandActions[command] || noop)();
onCommand(commandListener);
