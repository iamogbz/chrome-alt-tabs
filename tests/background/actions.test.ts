import { moveTabs, undo } from "background/actions";

describe("actions", () => {
    describe("move tabs", () => {
        const tabIds = [33, 22, 11];
        const mockTabs = tabIds.map(id => ({ id } as ChromeTab));

        it("creates valid action with no target", () => {
            expect(moveTabs({ tabs: mockTabs, from: 1 })).toMatchSnapshot();
        });

        it("creates valid action with no source", () => {
            expect(moveTabs({ tabs: mockTabs, to: 2 })).toMatchSnapshot();
        });

        it.each([null, undefined, []].map(v => [v]))(
            "fails to create invalid action for tabs:%s",
            async tabs => {
                expect(() => {
                    moveTabs({ tabs, from: 1, to: 2 });
                }).toThrowErrorMatchingSnapshot();
            },
        );
    });

    describe("undo", () => {
        it("creates valid action", () => {
            expect(undo()).toMatchSnapshot();
        });
    });
});
