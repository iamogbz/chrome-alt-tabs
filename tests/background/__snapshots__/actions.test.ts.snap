// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions move tabs creates valid action with no source 1`] = `
{
  "payload": {
    "from": undefined,
    "tabs": [
      {
        "id": 33,
      },
      {
        "id": 22,
      },
      {
        "id": 11,
      },
    ],
    "to": 2,
  },
  "type": "move-tabs",
}
`;

exports[`actions move tabs creates valid action with no target 1`] = `
{
  "payload": {
    "from": 1,
    "tabs": [
      {
        "id": 33,
      },
      {
        "id": 22,
      },
      {
        "id": 11,
      },
    ],
    "to": null,
  },
  "type": "move-tabs",
}
`;

exports[`actions move tabs fails to create invalid action for tabs:[] 1`] = `"Invalid Action Definition: move  (1 => 2)"`;

exports[`actions move tabs fails to create invalid action for tabs:null 1`] = `"Invalid Action Definition: move null (1 => 2)"`;

exports[`actions move tabs fails to create invalid action for tabs:undefined 1`] = `"Invalid Action Definition: move undefined (1 => 2)"`;

exports[`actions undo creates valid action 1`] = `
{
  "type": "undo",
}
`;
